### PR TITLE
feat: add firebase implementation of saved hikes repository

### PR DESCRIPTION
**Pull request by Wylwi**
# Summary
- The objective of this PR is to implement the real saved hikes repository, based on Firebase. 
- This PR contains code and tests to test the repository
- Removal of Dummy repository

# Details
- The firebase architecture is easy, having a savedHikes collections, which contains a document per user (the document id is the user uid). This document contains an array of SavedHikes
- I created a class named UserSavedHikes for making serializing and deserializing the document easier. 
- I defined also some new methods `getSavedHike` and `isHikeSaved` for helping to implement the View
- The repository also checks for a user to be defined, and throws an error if it isn't logged in
- I had to remove one test because it was specifically made for a dummy repository, and is no longer valid with Firebase

### CODE REVIEW

File: app/src/test/java/ch/hikemate/app/model/route/saved/SavedHikesRepositoryFirestoreTest.kt [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+    openMocks(this)
+
+    savedHikesRepositoryFirestore = SavedHikesRepositoryFirestore(mockFirestore, mockFirebaseAuth)
+
+    // If the collection is the one from the user, return the mock collection reference
+    `when`(mockFirestore.collection(eq(SavedHikesRepositoryFirestore.SAVED_HIKES_COLLECTION)))
+        .thenReturn(mockCollectionReference)
+    // If the collection is not the one from the user, throw an exception
+    `when`(mockFirestore.collection(not(eq(SavedHikesRepositoryFirestore.SAVED_HIKES_COLLECTION))))
+        .then { fail("Unexpected collection reference: ${it.arguments[0]}") }
```
[the remark this answers is not in the document]
Comment (Wylwi) [reply app/src/test/java/ch/hikemate/app/model/route/saved/SavedHikesRepositoryFirestoreTest.kt:58]: The comment is just no longer up to date! Thanks for the flag!

File: app/src/main/java/ch/hikemate/app/model/route/saved/SavedHikesRepository.kt [hunk 2]
```diff
@@ -35,6 +37,23 @@ interface SavedHikesRepository {
...
+   * @param id The ID of the saved hike to get.
+   * @return The saved hike with the provided ID, or null if it does not exist.
+   * @throws IllegalStateException If the user is not authenticated.
+   */
+  suspend fun getSavedHike(id: String): SavedHike?
+
+  /**
+   * Check if the hike with the provided ID exists in the saved hikes list.
+   *
+   * @param id The ID of the hike to check.
```
[the remark this answers is not in the document]
Comment (Wylwi) [reply app/src/main/java/ch/hikemate/app/model/route/saved/SavedHikesRepository.kt:56]: Great catch! Thank you!

File: app/src/main/java/ch/hikemate/app/model/route/saved/SavedHikesRepositoryFirestore.kt [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  override suspend fun isHikeSaved(id: String): Boolean {
+    checkNotNull(auth.currentUser) { ERROR_MSG_USER_NOT_AUTHENTICATED }
+    return db.collection(SAVED_HIKES_COLLECTION)
+        .document(auth.currentUser!!.uid)
+        .get()
+        .await()
+        .toObject(UserSavedHikes::class.java)
+        ?.savedHikes
+        ?.any { it.id == id } ?: false
+  }
```
[the remark this answers is not in the document]
Comment (Wylwi) [reply app/src/main/java/ch/hikemate/app/model/route/saved/SavedHikesRepositoryFirestore.kt]: Yep! Done!

File: app/src/main/java/ch/hikemate/app/model/route/saved/SavedHikesRepositoryFirestore.kt [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+  override suspend fun getSavedHike(id: String): SavedHike? {
+    checkNotNull(auth.currentUser) { ERROR_MSG_USER_NOT_AUTHENTICATED }
+    return db.collection(SAVED_HIKES_COLLECTION)
+        .document(auth.currentUser!!.uid)
+        .get()
+        .await()
+        .toObject(UserSavedHikes::class.java)
+        ?.savedHikes
+        ?.find { it.id == id }
+  }
```
[the remark this answers is not in the document]
Comment (Wylwi) [reply app/src/main/java/ch/hikemate/app/model/route/saved/SavedHikesRepositoryFirestore.kt]: Yep! Done!

**Review by rBergCode — APPROVED** [review]
Thanks for the changes. LGTM!

**Pull request merged**